doc/api/tracing.md

```diff
@@ -40,12 +67,125 @@ be specified with `--trace-event-file-pattern` that accepts a template
 string that supports `${rotation}` and `${pid}`. For example:
 
 ```txt
-node --trace-events-enabled --trace-event-file-pattern '${pid}-${rotation}.log' server.js
+node --trace-event-categories v8 --trace-event-file-pattern '${pid}-${rotation}.log' server.js
 ```
 
 Starting with Node.js 10.0.0, the tracing system uses the same time source
 as the one used by `process.hrtime()`
 however the trace-event timestamps are expressed in microseconds,
 unlike `process.hrtime()` which returns nanoseconds.
 
+## The `trace_events` module
+<!-- YAML
+added: REPLACEME
+-->
+
+### `Tracing` object
+<!-- YAML
+added: REPLACEME
+-->
+
+The `Tracing` object is used to enable or disable tracing for sets of
+categories. Instances are created using the `trace_events.createTracing()`
+method.
+
+#### `tracing.categories`
+<!-- YAML
+added: REPLACEME
+-->
+
+* {string}
+
+A comma-separated list of the trace event categories covered by this
+`Tracing` object.
+
+#### `tracing.disable()`
+<!-- YAML
+added: REPLACEME
+-->
+
+Disables this `Tracing` object.
```

Comment:
Do we also want to mention somewhere that the traced categories will get disabled if the `Tracing` object gets garbage collected. I wonder if this is going to be a gotcha for people – they might enable tracing then forget to keep a reference to the object long term?

Reply:
I don't really like that the categories are disabled if the `Tracing` object is garbage collected. We don't remove hooks in `async_hooks` just because the hooks object is removed. We don't stop a `setTimeout` just because the timer object is removed. I think it is perfectly natural to do:

```
trace_events.createTracing({ categories: ['node.http'] }).enable();
```

without ever disabling it again.

Reply:
In that case, however, it would be trivial to put this call into a context where it is never garbage collected, then you'd accomplish the same goal.

Reply:
> In that case, however, it would be trivial to put this call into a context where it is never garbage collected, then you'd accomplish the same goal.

I'm not really buying that as a good argument. You could say that about a lot of APIs we have. As a general rule, I don't think garbage collection should have side-effects.

Could we turn this discussion around, where you argue why it should have this side effect?

Reply:
> You could say that about a lot of APIs we have

That's fair ;-)

> Could we turn this discussion around, where you argue why it should have this side effect?

Essentially to ensure that the tracing can be turned off. For instance, without this, user code that *accidentally* loses the reference to the object would have no way of disabling the tracing for the process.

```js
{
  var tracing = trace_events.createTracing({ categories: ['node.http'] })
  tracing.enable()
}
// oops, can't disable any more ...
// creating and disabling a new tracing object won't help
{
  var tracing = trace_events.createTracing({ categories: ['node.http'] })
  tracing.disable()  // no impact, because it's a different tracing object
  // It is set up this way to avoid modules accidentally stepping on one another.
}
```

Reply:
I agree with @AndreasMadsen on this one. I think far more people are likely to fall into the first trap – dropping the reference to a `Tracing` object, and having tracing silently turn off at an arbitrary point in the future depending on when GC occurs.

I think it is quite intuitive in your example that calling `disable` on a new object isn't going to work. Users will understand this, and can easily fix their code – by keeping a reference to the `Tracing` instance.

Reply:
If tracing is enabled/disabled globally (single destination), should we do:


```js
trace_events.enable({ categories: ['node.http'] })
trace_events.disable({ categories: ['node.http'] })
trace_events.disable()
```

Reply:
@mcollina ... that's the approach my original PR on this took but the feedback was that having the `Tracing` objects to control it would be better.

Reply:
Removing the automatic disabling on collection would make it impossible to a developer to disable the tracing if they have lost a reference to that `tracing` object. If there is a `createTracing()`, then the tracing follows the scope of the object itself. Otherwise, there is a global resource. This mixture of scopes would be very hard to understand and a source of bugs.

@ofrobots @AndreasMadsen which one would you prefer: a) a module-level control to enable/disable or b) removing tracing when the tracing object goes out of scope?

Reply:
> Removing the automatic disabling on collection would make it impossible to a developer to disable the tracing if they have lost a reference to that `tracing` object.

This argument sounds so bad to me. Now two people are saying it, I feel there must be something fundamentally I have misunderstood.

I'm putting my foot down with this statement:

> If the user needs to disable tracing and they lose the `tracing` object, that is a bug in the user's code. Just like, if a user needs to clear a timeout and they lose the `timer` object, that is a bug in the user's code.

By disabling tracing on garbage collection we are working around buggy user code at the expense of a much more intuitive (read: no sudden side-effects) API.

Reply:
@AndreasMadsen given that tracing is global, how about we ditch createTracing completely, and just provide enable() and disable()?

Reply:
I would be concerned about this scenario:

```js
// module-1.js
trace_events.enable({ categories: ['node.http'] })
someTimeLater(() => {
  trace_events.disable({ categories: ['node.http'] }
})

// module-2.js
trace_events.enable({ categories: ['node.http'] })
```

Module-1 should not disable tracing globally. There is still someone else interested in tracing http. One way to mitigate this is by using a `createTracing`-like interface which becomes the handle that one can use to disable individual tracings. Alternatively, if we want a global API, then I think we should do reference counting on the categories. I would be okay with that. That probably also addresses this grumble from @jasnell:

> It makes me a bit twitchy allowing an object responsible for initiating the consumption of resources to be gc'd without cleaning up

Reply:
@jasnell I think that passing no category to `.disable()` should just disable everything independently of which categories are enabled.

Also, all operation (enabling-disabling) categories should emit a trace event. In this way whoever is listening can detect if the application has stopped emitting a certain category and warn the user.

Reply:
@jasnell @mcollina We have had this lengthy discussion before ([starts around here](https://github.com/nodejs/node/pull/19233#issuecomment-373510056)) and I thought we agreed that `trace_events.createTracing` API would be the best choice, as it allows for multiple actors and future proves the API for allowing an actor to specify an isolated output.

I couldn't find anywhere where you clearly argue why `trace_events.enable(categories)`/ `trace_events.disable(categories)` is better or how it solves the same use-cases. Could you highlight that for my blind eyes?

Reply:
@mcollina 

> However, by proposing that categories should not be disabled if the tracing object is gc'ed (in #19803 (comment)), you imply that tracing is a process concern.

Okay, let's get consensus if the `tracer` should be disabled on gc'ing first. The API discussion is meaningless to me before that.  Since we can't discuss gc behavior using the `trace_events.enable(categories)` API, let's consider the `trace_events.createTracing()` for now.


As I have stated before, I find the arguments completely misguided. I feel fundamentally stupid for not at least understanding where they come from. I think we need to take this discussion in a different direction if I'm to understand the logic behind these arguments. But before that, I will say that I would like:

```
trace_events.createTracing({
  categories: ['node.http'],
  destination: '{pid}.custom-trace-events'
}).enable()
```

to also work.

---

Alternative discussion strategy: Could we step away from `trace_events` and discuss 3 similar APIs and why they are this way.

* `timer = setTimeout()`
* `hooks = async_hooks.createHook()`
* `fd = new fs.Open()` (I couldn't find the PR, but I think I have heard @jasnell talk about this.)

Can we agree on the following being a good design?

* `timer = setTimeout()`: gc'ing `timer` **should not** call `clearTimeout`
* `hooks = async_hooks.createHook()`: gc'ing `hooks` **should not** call `hooks.disable()`
* `fd = new fs.Open()`: gc'ing `fd` **should** call `fs.close(fd)`.

Personally, I find the above good behavior.

The reason why I believe `fs.close(fd)` should be automatic, but the others shouldn't: is that if the user loses the reference to `fd` then the user can't call `fs.write(fd)` or any other fs API. There are thus no user consequences caused by calling `fs.close(fd)` automatically.

The same can not be said for `hooks` and `timer`. In those cases, the callbacks, and async hooks would not be called. This is directly observable by the users.

The same is the case for `tracer = trace_events.createTracer()`. If we call `tracer.disable()` automatically it is directly observable by the users. Not in-process but out-of-process in the `.trace-events` file.

This central question: "is cleanup directly observable by the users?", is why I find calling `tracer.disable()` on GC a really bad idea.

Reply:
I do think letting users lose a reference to `hooks` is less bad as this is. I would note that the hooks objects has no chance of being collected anyway, as it's typically referenced by the closures that are passed to the hook.

I agree that this matches the behavior of async_hooks.

`setTimeout`  is something that resolve itself. It does not leave any resource occupied after it's done, so it's not related.
I do agree that unused fd should be automatically closed.

I __keep disagreeing with this approach__, however let's land this and iterate separately, as it would need to be changed in async_hooks as well.

> This central question: "is cleanup directly observable by the users?", is why I find calling tracer.disable() on GC a really bad idea.

I find it as bad as allocating resources that could not be collected in the future.

Reply:
`setInterval()` is also a bit of a special case here because it's not going to be gc'd at all while the timer is running and the behavior with it is well established and expected with it by this point. If `setInterval()` were introduced today I'd be much more likely to advocate for more closely following the browser implementation that returns a number identifying the timer rather than an object, allowing `clearInterval()` to be called from any context. I'd also prefer to have an API that lets me discover which intervals are active. That ship has sailed for that particular API, however.

Another possible approach on this is to encourage folks to proactively disable the handles rather than wait for gc the same way we do with the new `FileHandle` class in `fs.promises`.  Essentially, emit a warning if the categories are disabled implicitly on gc rather than explicitly by calling the `disable()` method. It would be much more likely to encourage better API use while still cleaning up after ourselves.

Reply:
> If setInterval() were introduced today I'd be much more likely to advocate for more closely following the browser implementation that returns a number identifying the timer rather than an object, allowing clearInterval() to be called from any context.

What do we mean by `context`, maybe that is what I don't understand? How do you know what number to use?

Other examples:

* `child = child_process.spawn()` could call `child.kill()`.
* `server = http.createServer()` could call `server.close()`.
* `request = http.request()` could call `request.end()`.
* `stream = new zlib.Deflate()` could call `stream.close()`.
* `stream = new crypto.createHash()` could call `stream.end()`.